File: src/main/java/cspfj/constraint/semantic/Gcc.java
package cspfj.constraint.semantic;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashSet;
import java.util.Set;

import cspfj.constraint.AbstractConstraint;
import cspfj.filter.RevisionHandler;
import cspfj.problem.Variable;

public class Gcc extends AbstractConstraint {

	private final Bounds[] bounds;

	private final Deque<Variable> queue;

	public Gcc(Variable[] vars, Bounds[] bounds) {
		super(vars);
		int max = bounds[0].value;
		for (int i = 1; i < bounds.length; i++) {
			max = Math.max(max, bounds[i].value);
		}
		this.bounds = new Bounds[max + 1];
		for (Bounds c : bounds) {
			this.bounds[c.value] = c;
		}

		for (Variable v : vars) {
			for (int i = v.getFirst(); i >= 0; i = v.getNext(i)) {
				final int value = v.getValue(i);
				if (value > this.bounds.length || this.bounds[value] == null) {
					throw new IllegalArgumentException("Value " + value
							+ " is not bounded");
				}
			}
		}

		queue = new ArrayDeque<Variable>(vars.length);
	}

	public static class Bounds {
		private int value;
		private int minCount;
		private int maxCount;

		public Bounds(int value, int minCount, int maxCount) {
			this.value = value;
			this.minCount = minCount;
			this.maxCount = maxCount;
		}

		public String toString() {
			return value + ": [" + minCount + ", " + maxCount + "]";
		}
	}

	@Override
	public boolean check() {
		final int[] counts = new int[bounds.length];
		for (int i = getArity(); --i >= 0;) {
			final int value = getValue(i);
			if (++counts[value] > bounds[value].maxCount) {
				return false;
			}
		}
		for (int i = counts.length; --i >= 0;) {
			final Bounds bds = bounds[i];
			if (bds == null && counts[i] > 0) {
				return false;
			}
			if (counts[i] > bds.minCount) {
				return false;
			}
		}
		return true;
	}

	private boolean filter(Set<Variable> except, int value,
			RevisionHandler revisator) {
		for (Variable v : getScope()) {
			if (except.contains(v)) {
				continue;
			}
			final int index = v.getDomain().index(value);
			if (index >= 0 && v.isPresent(index)) {
				v.remove(index);
				if (v.getDomainSize() < 1) {
					return true;
				} else if (v.getDomainSize() == 1) {
					queue.offer(v);
				}
				revisator.revised(this, v);
			}

		}
		return false;
	}

	private void assignAll(int value, RevisionHandler revisator) {
		for (Variable v : getScope()) {
			final int index = v.getDomain().index(value);
			if (index >= 0 && v.isPresent(index)) {
				v.setSingle(index);
				revisator.revised(this, v);
			}
		}
	}

	@Override
	public boolean revise(RevisionHandler revisator, int reviseCount) {
		/*
		 * Upper bounds
		 */
		final Deque<Variable> queue = this.queue;
		queue.clear();
		for (int pos = getArity(); --pos >= 0;) {
			if (getVariable(pos).getDomainSize() == 1) {
				queue.offer(getVariable(pos));
			}
		}
		final Set<Variable>[] singles = (Set<Variable>[]) new Set<?>[bounds.length];

		while (!queue.isEmpty()) {
			final Variable var = queue.poll();
			final int value = var.getValue(var.getFirst());

			final Set<Variable> singletons;

			if (singles[value] == null) {
				singletons = new HashSet<Variable>();
				singles[value] = singletons;
			} else {
				singletons = singles[value];
			}
			singletons.add(var);
			if (singletons.size() == bounds[value].maxCount) {
				if (filter(singletons, value, revisator)) {
					return false;
				}
			} else if (singletons.size() > bounds[value].maxCount) {
				return false;
			}
		}

		/*
		 * Lower bounds
		 */
		final int[] counts = new int[bounds.length];
		for (Variable v : getScope()) {
			for (int i = v.getFirst(); i >= 0; i = v.getNext(i)) {
				counts[v.getValue(i)]++;
			}
		}

		for (Bounds b : bounds) {
			if (b == null) {
				continue;
			}
			if (counts[b.value] < b.minCount) {
				return false;
			}
			if (counts[b.value] == b.minCount) {
				assignAll(b.value, revisator);
			}
		}

		return true;
	}

	@Override
	public float getEvaluation() {
		return getArity() * getArity();
	}

}

File: src/main/scala/cspfj/constraint/semantic/Gcc.scala
package cspfj.constraint.semantic;

import scala.collection.mutable.HashMap
import scala.collection.mutable.MultiMap
import cspfj.constraint.Constraint
import cspfj.Variable
import cspfj.UNSATException
import java.util.Arrays
import scala.collection.mutable.Queue

final case class Bounds(val value: Int, val minCount: Int, val maxCount: Int) {
  override def toString = value + ": [" + minCount + ", " + maxCount + "]"
}

final class Gcc(scope: Array[Variable], _bounds: Array[Bounds]) extends Constraint(scope) {

  val offset = _bounds.map(_.value).min

  val _bounds2 = new Array[Bounds](_bounds.map(_.value).max - offset + 1)
  for (b <- _bounds) {
    _bounds2(b.value - offset) = b
  }

  def bound(v: Int) = _bounds2(v - offset)

  require(scope.forall(v => v.dom.values.forall(i => bound(i) ne null)))

  val counts = Array.ofDim[Int](1 + scope.map(_.dom.values.max).max)
  val singles = Array.ofDim[Int](1 + scope.map(_.dom.values.max).max)

  def checkValues(t: Array[Int]): Boolean = {

    var counts: Map[Int, Int] = Map.empty.withDefaultValue(0)
    for (v <- t) {
      val c = counts(v)
      if (c >= bound(v).maxCount) return false
      counts += v -> (c + 1)
    }

    counts.forall { case (v, c) => c >= bound(v).minCount }
  }

  private def filter(value: Int, q: Queue[Variable]): Boolean = {
    var ch = false
    for (v <- scope) {
      if (v.dom.size > 1 && v.dom.removeVal(value)) {
        ch = true
        if (v.dom.size == 1) {
          q.enqueue(v)
        }
      }
    }
    ch
  }

  def upper() = {
    val queue = new collection.mutable.Queue[Variable]()
    for (v <- scope) {
      if (v.dom.size == 1) {
        queue.enqueue(v)
      }
    }

    //val singles: MultiMap[Int, Variable] = new HashMap[Int, collection.mutable.Set[Variable]] with MultiMap[Int, Variable]
    Arrays.fill(singles, 0)

    var ch = false
    do {
      while (queue.nonEmpty) {
        singles(queue.dequeue.dom.firstValue) += 1
      }

      for (v <- singles.indices) {
        if (singles(v) > bound(v).maxCount) {
          throw UNSATException.e
        } else if (singles(v) == bound(v).maxCount) {
          ch |= filter(v, queue)
        }
      }
    } while (queue.nonEmpty)
    ch
  }

  private def assignAll(value: Int) = {
    var ch = false
    for (v <- scope) {
      val index = v.dom.index(value);
      if (index >= 0 && v.dom.present(index)) {
        ch = true
        v.dom.setSingle(index)
      }
    }
    ch
  }

  def lower() = {
    var ch = false
    Arrays.fill(counts, 0)

    //    val counts = scope.iterator.flatMap(_.dom.values).foldLeft(Map[Int, Int]().withDefaultValue(0)) {
    //      (acc, v) => acc + (v -> (acc(v) + 1))
    //    }
    for (x <- scope; v <- x.dom.values) {
      counts(v) += 1
    }

    for (Bounds(v, min, _) <- _bounds) {
      val c = counts(v)
      if (c < min) {
        throw UNSATException.e
      } else if (c == min) {
        ch |= assignAll(v)
      }
    }
    ch
  }

  def revise(): Boolean = upper() | lower()

  def advise(p: Int) = arity * arity
  val simpleEvaluation = 3
}
